### PR TITLE
8013527: calling MethodHandles.lookup on itself leads to errors

### PR DESCRIPTION
**Pull request by mlchung**
JDK-8013527: calling MethodHandles.lookup on itself leads to errors
JDK-8257874: MethodHandle injected invoker doesn't have necessary private access

Johannes Kuhn is also a contributor to this patch.

A caller-sensitive method can behave differently depending on the identity
of its immediate caller.  If a method handle for a caller-sensitive method is
requested, this resulting method handle behaves as if it were called from an
instruction contained in the lookup class.  The current implementation injects
a trampoline class (aka the invoker class) which is the caller class invoking
such caller-sensitive method handle.  It works in all CSMs except `MethodHandles::lookup`
because the caller-sensitive behavior depends on the module of the caller class,
the class loader of the caller class, the accessibility of the caller class, or
the protection domain of the caller class.  The invoker class is a hidden class
defined in the same runtime package with the same protection domain as the
lookup class, which is why the current implementation works for all CSMs except
`MethodHandles::lookup` which uses the caller class as the lookup class.

Two issues with current implementation:
1.  The invoker class only has the package access as the lookup class.  It cannot
    access private members of the lookup class and its nest members.

    The fix is to make the invoker class as a nestmate of the lookup class.

2.  `MethodHandles::lookup` if invoked via a method handle produces a `Lookup`
    object of an injected invoker class which is a bug.

There are two alternatives:
- define the invoker class with the lookup class as the class data such that
  `MethodHandles::lookup` will get the caller class from the class data if
  it's the injected invoker
- Johannes' proposal [1]: detect if an alternate implementation with an additional
  trailing caller class parameter is present, use the alternate implementation
  and bind the method handle with the lookup class as the caller class argument.

There has been several discussions on the improvement to support caller sensitive
methods for example the calling sequences and security implication.   I have
looked at how each CSM uses the caller class.  The second approach (i.e. 
defining an alternate implementation for a caller-sensitive method taking
an additional caller class parameter) does not work for non-static non-final
caller-sensitive method.  In addition, it is not ideal to pollute the source
code to provide an alternatve implementation for all 120+ caller-sensitive methods
whereas the injected invoker works for all except `MethodHandles::lookup`.

I propose to use both approaches.   We can add an alternative implementation for
a caller-sensitive method when desirable such as `MethodHandles::lookup` in
this PR.  For the injected invoker case, one could extract the original lookup
class from class data if needed.

test/jdk/jdk/internal/reflect/CallerSensitive/CheckCSM.java ensures that
no new non-static non-final caller-sensitive method will be added to the JDK.
I extend this test to catch that non-static non-final caller-sensitive method
cannot have the alternate implementation taking the additional caller class
parameter.

This fix for JDK-8013527 is needed by the prototype for JDK-6824466 I'm working on.

[1] https://mail.openjdk.java.net/pipermail/core-libs-dev/2021-January/073184.html

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issues
 * [JDK-8013527](https://bugs.openjdk.java.net/browse/JDK-8013527): calling MethodHandles.lookup on itself leads to errors
 * [JDK-8257874](https://bugs.openjdk.java.net/browse/JDK-8257874): MethodHandle injected invoker doesn't have necessary private access

### Reviewers
 * [Johannes Kuhn](https://openjdk.java.net/census#jkuhn) (@DasBrain - Author)

### Contributors
 * Johannes Kuhn `<info@j-kuhn.de>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2367/head:pull/2367`
`$ git checkout pull/2367`

### CODE REVIEW

**Review by DasBrain — CHANGES_REQUESTED** [review]
Thanks Mandy.

Looks good, except the possibility for an attacker to teleport within the same nest.

File: src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java [hunk 1]
```diff
@@ -1142,39 +1145,101 @@ static MethodHandle bindCaller(MethodHandle mh, Class<?> hostClass) {
...
+             * @CSM must be public and exported if called by any module.
+             */
+            String name = targetClass.getName() + INVOKER_SUFFIX;
+            if (targetClass.isHidden()) {
+                // use the original class name
+                name = name.replace('/', '_');
+            }
+            Class<?> invokerClass = new Lookup(targetClass)
+                    .makeHiddenClassDefiner(name, INJECTED_INVOKER_TEMPLATE, Set.of(NESTMATE))
+                    .defineClass(true, targetClass);
```
[the remark this answers is not in the document]
Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:1205]: Only `Lookup` with the original access can access `MethodHandles.classData`.   A hidden class `HC$$InjectedInvoker/0x1234` can access private members of another class `C` in the same nest but not `C`'s class data.

[the remark this answers is not in the document]
Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:1205]: Last night, I had a second thought that the fix for these two JBS issues does not need the class data.  It's more for a future use.  I plan to revise it and drop class data in this fix.

[the remark this answers is not in the document]
Comment (DasBrain) [reply src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:1205]: You are right, got it confused with the future use.

With this fix, MethodHandle -> Method.invoke -> MethodHandles.lookup() will still return a lookup on the injected invoker.  
I somehow missed that this is not part of the fix, but for the future use.

[the remark this answers is not in the document]
Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:1205]: `MethodHandle -> Method.invoke -> MethodHandles.lookup() ` is a corner case that can be fixed easily using the class data approach.  See the new commit.

[the remark this answers is not in the document]
Comment (DasBrain) [reply src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:1235]: The security issue I mentioned was in an other branch, method-invoke.  

I used commit https://github.com/mlchung/jdk/commit/4a3c914f1b46cf84b42f6b6bc19d421955faac3f (i.e. before strengthening the injected invoker checks) to test the [my exploit](https://gist.github.com/DasBrain/4dda6cc3a13e1636afe17e6a02ec3d12). (Yes, full sandbox escape.)

I hope the same is not possible with the nestmate requirement.  

PS.: Hidden Class -> MethodHandle -> Method.invoke -> MethodHandles might break due to mangling of the hidden class name for the injected invoker. Will write a test.

[the remark this answers is not in the document]
Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:1235]: Well, my branch `method-invoke` is a prototype and work-in-progress.   I won't dig too much to it.

Do you reproduce the issue with this patch?   The fix will ensure it's the invoker class injected by BindCaller.

**Review by DasBrain — APPROVED** [review]
Looks good.

**Pull request closed without merging**